cyfra-spirv-tools/src/main/scala/io/computenode/cyfra/spirvtools/SpirvTool.scala

```diff
@@ -0,0 +1,140 @@
+      def loopOverBuffer(buf: Array[Byte]): Unit = {
+        val len = inStream.read(buf)
+        if (len == -1) ()
+        else {
+          outStream.write(buf, 0, len)
+          loopOverBuffer(buf)
+        }
+      }
+
+      Using
+        .Manager { use =>
+          val in = use(inStream)
+          val out = use(outStream)
+          val buf = new Array[Byte](1024)
+          loopOverBuffer(buf)
+          out.flush()
+        }
+        .toEither
+        .left
+        .map(e => SpirvToolIOError(s"$description failed: ${e.getMessage}"))
+    }
+
+    def createProcessIO(): Either[SpirvError, ProcessIO] = {
+      val inHandler: OutputStream => Unit =
+        in =>
+          safeIOCopy(inputStream, in, "Writing to stdin") match {
+            case Left(err) => SpirvToolIOError(s"Failed to create ProcessIO: ${err.getMessage}")
+            case Right(_)  => ()
+          }
+
+      val outHandler: InputStream => Unit =
+        out =>
+          safeIOCopy(out, outputStream, "Reading stdout") match {
+            case Left(err) => SpirvToolIOError(s"Failed to create ProcessIO: ${err.getMessage}")
+            case Right(_)  => ()
+          }
+
+      val errHandler: InputStream => Unit =
+        err =>
+          safeIOCopy(err, errorStream, "Reading stderr") match {
+            case Left(err) => SpirvToolIOError(s"Failed to create ProcessIO: ${err.getMessage}")
+            case Right(_)  => ()
+          }
+
+      Try {
+        new ProcessIO(inHandler, outHandler, errHandler)
+      }.toEither.left.map(e => SpirvToolIOError(s"Failed to create ProcessIO: ${e.getMessage}"))
+    }
+
+    for {
+      processIO <- createProcessIO()
+      process <- Try(cmd.run(processIO)).toEither.left.map(ex => SpirvCommandExecutionFailed(s"Failed to execute SPIR-V command: ${ex.getMessage}"))
+    } yield (outputStream, errorStream, process.exitValue())
+  }
+
+  trait SpirvError extends RuntimeException {
+    def message: String
+
+    override def getMessage: String = message
+  }
+
+  protected trait ResultSaveSetting
+
+  case class ToFile(filePath: Path) extends ResultSaveSetting {
+    require(filePath != null, "filePath must not be null")
+    Option(filePath.getParent).foreach { dir =>
+      if (!Files.exists(dir)) {
+        Files.createDirectories(dir)
+        logger.debug(s"Created output directory: $dir")
+      }
+    }
+  }
+
+  final case class SpirvToolNotFound(toolName: String) extends SpirvError {
+    def message: String = s"Tool '$toolName' not found in PATH."
+  }
+
+  final case class SpirvCommandExecutionFailed(details: String) extends SpirvError {
+    def message: String = s"SPIR-V command execution failed: $details"
+  }
+
+  final case class SpirvToolIOError(details: String) extends SpirvError {
+    def message: String = s"SPIR-V command encountered IO error: $details"
+  }
+
+  case object NoSaving extends ResultSaveSetting
+}
+
+object SpirvTool {
+  def dumpSpvToFile(code: ByteBuffer, path: Path): Unit = {
```

Comment:
It can be universal, it just dumps bytes to a file, not necessarily spirv. This could end up in the ToFile class and support also text in one spot.